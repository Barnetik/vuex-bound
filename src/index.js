export const normalizeMap = map => (
  Array.isArray(map)
    ? map.map(key => ({ key, val: key }))
    : Object.keys(map).map(key => ({ key, val: map[key] }))
);

export const normalizeNamespace = func =>
  (namespace, map) => {
    if (typeof namespace !== 'string') {
      map = namespace;
      namespace = '';
    }

    return func(namespace, map);
  };

export const mapModel = normalizeNamespace((namespace, models) => {
  const res = {};

  normalizeMap(models).forEach(({ key, val }) => {
    res[key] = {
      get() {
        if (typeof val === 'function') val(this.$store.state);

        if (!namespace) return this.$store.state[key];

        return namespace.split('/')
          .reduce((prev, cur) => prev[cur], this.$store.state)[key];
      },
      set(value) {
        const type = !namespace ? 'updateModel' : `${namespace.split('/').join('/')}/updateModel`;

        const valify = String(val);
        const originState = valify.substring(valify.indexOf('.') + 1, valify.indexOf(';'));

        this.$store.commit(type, { label: originState || key, value });
      },
    };
  });

  return res;
});

export const updateModel = () => ({
  updateModel(state, { label, value }) {
    state[label] = value;
  },
});
