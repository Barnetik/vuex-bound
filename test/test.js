import { mapModelsToState, updateModel } from '../src';

process.chdir(__dirname);

describe('vuex-bound', () => {
  it('should handle mapModelsToState for module name', () => {
    const MODULE_NAME = 'foo';
    const computed = { ...mapModelsToState(MODULE_NAME, ['modelName1', 'modelName2']) };

    expect(computed).toMatchSnapshot();
  });

  it('should handle mapModelsToState for sub-module name', () => {
    const MODULE_NAME = 'foo.bar';
    const computed = { ...mapModelsToState(MODULE_NAME, ['modelName1', 'modelName2']) };

    expect(computed).toMatchSnapshot();
  });

  it('should handle mapModelsToState for sub-sub-module name', () => {
    const MODULE_NAME = 'foo.bar.baz';
    const computed = { ...mapModelsToState(MODULE_NAME, ['modelName1', 'modelName2', 'modelName3']) };

    expect(computed).toMatchSnapshot();
  });

  it('should handle updateModel for module name', () => {
    const state = {
      field: 'name',
    };

    const payload = {
      label: 'field',
      value: 'change',
    };

    updateModel().update(state, payload);

    expect(state.field).toBe('change');
  });
});
