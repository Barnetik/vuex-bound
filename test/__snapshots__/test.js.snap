// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`vuex-bound should handle mapModelsToState for module name 1`] = `
Object {
  "modelName1": Object {
    "get": [Function],
    "set": [Function],
  },
  "modelName2": Object {
    "get": [Function],
    "set": [Function],
  },
}
`;

exports[`vuex-bound should handle mapModelsToState for sub-module name 1`] = `
Object {
  "modelName1": Object {
    "get": [Function],
    "set": [Function],
  },
  "modelName2": Object {
    "get": [Function],
    "set": [Function],
  },
}
`;

exports[`vuex-bound should handle mapModelsToState for sub-sub-module name 1`] = `
Object {
  "modelName1": Object {
    "get": [Function],
    "set": [Function],
  },
  "modelName2": Object {
    "get": [Function],
    "set": [Function],
  },
  "modelName3": Object {
    "get": [Function],
    "set": [Function],
  },
}
`;
